crates/red_knot_python_semantic/src/types.rs

```diff
@@ -1359,193 +1407,449 @@ impl<'db> Type<'db> {
-            },
+            let descriptor_kind = if self.class_member(db, "__set__".into()).symbol.is_unbound()
+                && self
+                    .class_member(db, "__delete__".into())
+                    .symbol
+                    .is_unbound()
+            {
+                AttributeKind::NormalOrNonDataDescriptor
+            } else {
+                AttributeKind::DataDescriptor
+            };
 
-            Type::StringLiteral(_) | Type::LiteralString => {
-                KnownClass::Str.to_instance(db).static_member(db, name)
-            }
+            Some((return_ty, descriptor_kind))
+        } else {
+            None
+        }
+    }
 
-            Type::BytesLiteral(_) => KnownClass::Bytes.to_instance(db).static_member(db, name),
+    /// Look up `__get__` on the meta type of `attribute`, and call it with `attribute`, `instance`,
+    /// and `owner` as arguments. This method exists as a separate step as we need to handle unions
+    /// and intersections explicitly.
+    fn try_call_dunder_get_on_attribute(
+        db: &'db dyn Db,
+        attribute: SymbolAndQualifiers<'db>,
+        instance: Type<'db>,
+        owner: Type<'db>,
+    ) -> (SymbolAndQualifiers<'db>, AttributeKind) {
+        match attribute {
+            // This branch is not strictly needed, but it short-circuits the lookup of various dunder
+            // methods and calls that would otherwise be made.
+            //
+            // Note that attribute accesses on dynamic types always succeed. For this reason, they also
+            // have `__get__`, `__set__`, and `__delete__` methods and are therefore considered to be
+            // data descriptors.
+            //
+            // The same is true for `Never`.
+            SymbolAndQualifiers {
+                symbol: Symbol::Type(Type::Dynamic(_) | Type::Never, _),
+                qualifiers: _,
+            } => (attribute, AttributeKind::DataDescriptor),
+
+            SymbolAndQualifiers {
+                symbol: Symbol::Type(Type::Union(union), boundness),
+                qualifiers,
+            } => (
+                union
+                    .map_with_boundness(db, |elem| {
+                        Symbol::Type(
+                            elem.try_call_dunder_get(db, instance, owner)
+                                .map_or(*elem, |(ty, _)| ty),
+                            boundness,
+                        )
+                    })
+                    .with_qualifiers(qualifiers),
+                // TODO: avoid the duplication here:
+                if union.elements(db).iter().all(|elem| {
+                    elem.try_call_dunder_get(db, instance, owner)
+                        .is_some_and(|(_, kind)| kind.is_data())
+                }) {
+                    AttributeKind::DataDescriptor
+                } else {
+                    AttributeKind::NormalOrNonDataDescriptor
+                },
```

Comment:
Iterating over the union elements twice here just to be able to use `map_with_boundness` is not great. I think we might need a more generic `map`/`fold` function on unions and intersections. We often have to "union" other meta data as well, not just boundness, but also type qualifiers or above: whether or not a symbol is considered a data descriptor.

Reply:
I will also leave this for a follow up